beacon/light/request/scheduler.go

```diff
@@ -0,0 +1,248 @@
+		// that after a trigger happens testWaitCh will block until the resulting
+		// processing round has been finished
+		triggerCh: make(chan struct{}, 1),
+		//testWaitCh: make(chan struct{}),
+	}
+	return s
+}
+
+// RegisterModule registers a module. Should be called before starting the scheduler.
+// In each processing round the order of module processing depends on the order of
+// registration.
+func (s *Scheduler) RegisterModule(m Module) {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	s.modules = append(s.modules, m)
+}
+
+// RegisterServer registers a new server.
+func (s *Scheduler) RegisterServer(server Server) {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	s.handleEvent(server, Event{Type: EvRegistered})
+	server.Subscribe(func(event Event) {
+		s.lock.Lock()
+		if _, ok := s.servers[server]; ok {
+			s.handleEvent(server, event)
+		} else {
+			log.Error("Event received from unsubscribed server")
+		}
+		s.lock.Unlock()
+	})
+	s.servers[server] = struct{}{}
+}
+
+// UnregisterServer removes a registered server.
+func (s *Scheduler) UnregisterServer(server Server) {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	server.Unsubscribe()
+	delete(s.servers, server)
+	s.handleEvent(server, Event{Type: EvUnregistered})
+}
+
+// Start starts the scheduler. It should be called after registering all modules
+// and before registering any servers.
+func (s *Scheduler) Start() {
+	go s.syncLoop()
+}
+
+// Stop stops the scheduler.
+func (s *Scheduler) Stop() {
+	s.lock.Lock()
+	for server, _ := range s.servers {
+		server.Unsubscribe()
+	}
+	s.servers = nil
+	s.lock.Unlock()
+	stop := make(chan struct{})
+	s.stopCh <- stop
+	<-stop
+}
+
+// syncLoop calls all processable modules in the order of their registration.
+// A round of processing starts whenever there is at least one processable module.
+// Triggers triggered during a processing round do not affect the current round
+// but ensure that there is going to be a next round.
+func (s *Scheduler) syncLoop() {
+	for {
+		s.lock.Lock()
+		s.processModules()
+		s.lock.Unlock()
+
+	loop:
+		for {
+			select {
+			case stop := <-s.stopCh:
+				close(stop)
+				return
+			case <-s.triggerCh:
+				break loop
+				//case <-s.testWaitCh:
+			}
+		}
+	}
+}
+
+// processModules runs an entire processing round, calling processable modules
+// with the appropriate Environment.
+func (s *Scheduler) processModules() {
+	servers := make(serverSet)
+	for server, _ := range s.servers {
+		if ok, _ := server.CanRequestNow(); ok {
+			servers[server] = struct{}{}
+		}
+	}
+	for _, module := range s.modules {
+		tracker := s.trackers[module]
+		tracker.servers = servers
+		if module.Process(tracker, tracker.requestEvents, s.serverEvents) {
+			s.Trigger()
+		}
+		tracker.requestEvents = nil
+	}
+	s.serverEvents = nil
+}
+
+func (s *Scheduler) Trigger() {
+	select {
+	case s.triggerCh <- struct{}{}:
+	default:
+	}
+}
+
+func (s *Scheduler) handleEvent(server Server, event Event) {
+	s.Trigger()
```

Comment:
Not 100% sure, but if we trigger the loop here, wouldn't we run into a race where the modules would be triggered with old serverEvents?
Afaik we should handle the events, add the serverEvents and then trigger the modules

Reply:
I think it's correct, this happens under mutex lock and the next process round can only look at the event lists after this function is finished. Maybe it would look more intuitive though if I deferred the trigger after executing the function.